### ISSUE TITLE
htop explained

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#5585]
htop explained<br>
<br>
https://ift.tt/2gTZcty<br>
<br>
P&#275;teris &#325;ikiforovs<br>
<br>
<div>
<div>
<p>For the longest time I did not know what everything meant in htop.</p>
<p>I thought that load average <code>1.0</code> on my two core machine means that the CPU usage is at 50%. That's not quite right. And also, why does it say <code>1.0</code>?</p>
<p>I decided to look everything up and document it here.</p>
<p>They also say that the best way to learn something is to try to teach it.</p>
<p>Table of Contents</p>
<p>htop on Ubuntu Server 16.04 x64</p>
<p>Here is a screenshot of htop that I am going to describe.</p>
<p><img alt="Screenshot of htop" src="https://peteris.rocks/blog/htop/canyoukillit-before.png"></p>
<p>Uptime</p>
<p>Uptime shows how long the system has been running.</p>
<p>You can see the same information by running <code>uptime</code>:</p>
<pre><code>$ uptime 12:17:58 up 111 days, 31 min, 1 user, load average: 0.00, 0.01, 0.05 </code></pre>
<p>How does the <code>uptime</code> program know that?</p>
<p>It reads the information from the file <code>/proc/uptime</code>.</p>
<pre><code>9592411.58 9566042.33 </code></pre>
<p>The first number is the total number of seconds the system has been up. The second number is how much of that time the machine has spent idle, in seconds The second value may be greater than the overall system uptime on systems with multiple cores since it is a sum.</p>
<p>How did I know that? I looked at what files the <code>uptime</code> program opens when it is run. We can use the <code>strace</code> tool to do that.</p>
<pre><code>strace uptime </code></pre>
<p>There will be a lot of output. We can <code>grep</code> for the <code>open</code> system call. But that will not really work since <code>strace</code> outputs everything to the standard error (stderr) stream. We can redirect the stderr to the standard output (stdout) stream with <code>2&gt;&amp;1</code>.</p>
<p>Our output is this:</p>
<pre><code>$ strace uptime 2&gt;&amp;1 | grep open ... open(<span>"/proc/uptime"</span>, O_RDONLY) = 3 open(<span>"/var/run/utmp"</span>, O_RDONLY|O_CLOEXEC) = 4 open(<span>"/proc/loadavg"</span>, O_RDONLY) = 4 </code></pre>
<p>which contains the file <code>/proc/uptime</code> which I mentioned.</p>
<p>It turns out that you can also use <code>strace -e open uptime</code> and not bother with grepping.</p>
<p>So why do we need the <code>uptime</code> program if we can just read the contents of the file? The <code>uptime</code> output is nicely formatted for humans whereas the number of seconds is more useful for using in your own programs or scripts.</p>
<p>Load average</p>
<p>In addition to uptime, there were also three numbers that represent the load average.</p>
<pre><code>$ uptime 12:59:09 up 32 min, 1 user, load average: 0.00, 0.01, 0.03 </code></pre>
<p>They are taken from the <code>/proc/loadavg</code> file. If you take another look at the <code>strace</code> output, you'll see that this file was also opened.</p>
<pre><code>$ cat /proc/loadavg 0.00 0.01 0.03 1/120 1500 </code></pre>
<p>The first three columns represent the average system load of the last 1, 5, and 15 minute periods. The fourth column shows the number of currently running processes and the total number of processes. The last column displays the last process ID used.</p>
<p>Let's start with the last number.</p>
<p>Whenever you launch a new process, it is assigned an ID number. Process IDs are usually increasing, unless they've been exausted and are being reused. The process ID of 1 belongs to <code>/sbin/init</code> which is started at boot time.</p>
<p>Let's look at the <code>/proc/loadavg</code> contents again and then launch the <code>sleep</code> command in the background. When it's launched in the background, its process ID will be shown.</p>
<pre><code>$ cat /proc/loadavg 0.00 0.01 0.03 1/123 1566 $ sleep 10 &amp; [1] 1567 </code></pre>
<p>So the <code>1/123</code> means that there is one process running or ready to run at this time and there are <code>123</code> processed in total.</p>
<p>When you run <code>htop</code> and see just one running process, it means that it is the <code>htop</code> process itself.</p>
<p>If you run <code>sleep 30</code> and run <code>htop</code> again, you'll notice that there is still just 1 running process. That's because <code>sleep</code> is not running, it is sleeping or idling or in other words waiting for something to happen. A running process is a process that is currently running on the physical CPU or waiting its turn to run on the CPU.</p>
<p>If you run <code>cat /dev/urandom &gt; /dev/null</code> which repeatedly generates random bytes and writes them to a special file that is never read from, you will see that there are now two running process.</p>
<pre><code>$ cat /dev/urandom &gt; /dev/null &amp; [1] 1639 $ cat /proc/loadavg 1.00 0.69 0.35 2/124 1679 </code></pre>
<p>So there are now two running processes (random number generation and the <code>cat</code> that reads the contents of <code>/proc/loadavg</code>) and you'll also notice that the load averages have increased.</p>
<p>The load average represents the average system load over a period of time.</p>
<p>The load number is calculated by counting the number of running (currently running or waiting to run) and uninterruptible processes (waiting for disk or network activity). So it's simply a number of processes.</p>
<p>The load averages are then the average number of those processes during the last 1, 5 and 15 minutes, right?</p>
<p>It turns out it's not as simple as that.</p>
<p>The load average is the exponentially damped moving average of the load number. From Wikipedia:</p>
<blockquote>
<p>Mathematically speaking, all three values always average all the system load since the system started up. They all decay exponentially, but they decay at different speed. Hence, the 1-minute load average will add up 63% of the load from last minute, plus 37% of the load since start up excluding the last minute. Therefore, it's not technically accurate that the 1-minute load average only includes the last 60 seconds activity (since it still includes 37% activity from the past), but that includes mostly the last minute.</p>
</blockquote>
<p>Is that what you expected?</p>
<p>Let's return to our random number generation.</p>
<pre><code>$ cat /proc/loadavg 1.00 0.69 0.35 2/124 1679 </code></pre>
<p>While technically not correct, this is how I simplify load averages to make it easier to reason about them.</p>
<p>In this case, the random number generation process is CPU bound, so the load average over the last minute is <code>1.00</code> or on average 1 running process.</p>
<p>Since there is only one CPU on my system, the CPU utilization is 100% since my CPU can run only one process at a time.</p>
<p>If I had two cores, my CPU usage would be 50% since my computer can run two processes at the same time. The load average of a computer with 2 cores that has a 100% CPU utilization would be <code>2.00</code>.</p>
<p>You can see the number of your cores or CPUs in the top left corner of <code>htop</code> or by running <code>nproc</code>.</p>
<p>Because the load number also includes processes in uninterruptible states which don't have much effect on CPU utilization, it's not quite correct to infer CPU usage from load averages like I just did. This also explains why you may see high load averages but not much load on the CPU.</p>
<p>But there are tools like <code>mpstat</code> that can show the instantaneous CPU utilization.</p>
<pre><code>$ sudo apt install sysstat -y $ mpstat 1 Linux 4.4.0-47-generic (hostname) 12/03/2016 _x86_64_ (1 CPU) 10:16:20 PM CPU %usr %nice %sys %iowait %irq %soft %steal %guest %gnice %idle 10:16:21 PM all 0.00 0.00 100.00 0.00 0.00 0.00 0.00 0.00 0.00 0.00 10:16:22 PM all 0.00 0.00 100.00 0.00 0.00 0.00 0.00 0.00 0.00 0.00 10:16:23 PM all 0.00 0.00 100.00 0.00 0.00 0.00 0.00 0.00 0.00 0.00 <span># ...</span> <span># kill cat /dev/urandom</span> <span># ...</span> 10:17:00 PM all 0.00 0.00 0.00 0.00 0.00 0.00 0.00 0.00 0.00 100.00 10:17:01 PM all 1.00 0.00 0.00 2.00 0.00 0.00 0.00 0.00 0.00 97.00 10:17:02 PM all 0.00 0.00 0.00 0.00 0.00 0.00 0.00 0.00 0.00 100.00 </code></pre>
<p>Why do we use load averages then?</p>
<pre><code>$ curl -s https://raw.githubusercontent.com/torvalds/linux/v4.8/kernel/sched/loadavg.c | head -n 7 /* * kernel/sched/loadavg.c * * This file contains the magic bits required to compute the global loadavg * figure. Its a silly number but people think its important. We go through * great pains to make it work on big machines and tickless kernels. */ </code></pre>
<p>Processes</p>
<p>In the top right corner, <code>htop</code> shows the total number of processes and how many of them are running. But it says <em>Tasks</em> not processes. Why?</p>
<p>Another name for a process is a <em>task</em>. The Linux kernel internally refers to processes as tasks. <code>htop</code> uses Tasks instead of Processes probably because it's shorter and saves some screen space.</p>
<p>You can also see threads in <code>htop</code>. To toggle the visibility of threads, hit <code>Shift</code>+<code>H</code> on your keyboard. If you see <code>Tasks: 23, 10 thr</code>, it means it they are visible.</p>
<p>You can also see kernel threads with <code>Shift</code>+<code>K</code>. When they are visible, it'll say <code>Tasks: 23, 40 kthr</code>.</p>
<p>Process ID / PID</p>
<p>Every time a new process is started it is assigned an identification number (ID) which is called process ID or PID for short.</p>
<p>If you run a program in the background (<code>&amp;</code>) from <code>bash</code>, you will see the job number in square brackets and the PID.</p>
<pre><code>$ sleep 1000 &amp; [1] 12503 </code></pre>
<p>If you missed it, you can use the <code>$!</code> variable in <code>bash</code> that will expand to the last backgrounded process ID.</p>
<pre><code>$ <span>echo</span> $! 12503 </code></pre>
<p>Process ID is very useful. It can be used to see details about the process and to control it.</p>
<p><code>procfs</code> is a pseudo file system that lets userland programs to get information from the kernel by reading files. It is usually mounted at <code>/proc/</code> and to you it looks like a regular directory that you can browse with <code>ls</code> and <code>cd</code>.</p>
<p>All information related to a process is located at <code>/proc/&lt;pid&gt;/</code>.</p>
<pre><code>$ ls /proc/12503 attr coredump_filter fdinfo maps ns personality smaps task auxv cpuset gid_map mem numa_maps projid_map stack uid_map cgroup cwd io mountinfo oom_adj root <span>stat</span> wchan clear_refs environ limits mounts oom_score schedstat statm cmdline exe loginuid mountstats oom_score_adj sessionid status comm fd map_files net pagemap setgroups syscall </code></pre>
<p>For example, <code>/proc/&lt;pid&gt;/cmdline</code> will give the command that was used to launch the process.</p>
<pre><code>$ cat /proc/12503/cmdline sleep1000$ </code></pre>
<p>Ugh, that's not right. It turns out that the command is separated by the <code>\0</code> byte.</p>
<pre><code>$ od -c /proc/12503/cmdline 0000000 s l e e p \0 1 0 0 0 \0 0000013 </code></pre>
<p>which we can replace with a space or newline</p>
<pre><code>$ tr <span>'\0'</span> <span>'\n'</span> &lt; /proc/12503/cmdline sleep 1000 $ strings /proc/12503/cmdline sleep 1000 </code></pre>
<p>The process directory for a process can contain links! For instance, <code>cwd</code> points to the current working directory and <code>exe</code> is the executed binary.</p>
<pre><code>$ ls <span>-l</span> /proc/12503/{cwd,exe} lrwxrwxrwx 1 ubuntu ubuntu 0 Jul 6 10:10 /proc/12503/cwd -&gt; /home/ubuntu lrwxrwxrwx 1 ubuntu ubuntu 0 Jul 6 10:10 /proc/12503/exe -&gt; /bin/sleep </code></pre>
<p>So this is how <code>htop</code>, <code>top</code>, <code>ps</code> and other diagnostic utilities get their information about the details of a process: they read it from <code>/proc/&lt;pid&gt;/&lt;file&gt;</code>.</p>
<p>Process tree</p>
<p>When you launch a new process, the process that launched the new process is called the parent process. The new process is now a child process for the parent process. These relationships form a tree structure.</p>
<p>If you hit <code>F5</code> in <code>htop</code>, you can see the process hierarchy.</p>
<p>You can also use the <code>f</code> switch with <code>ps</code></p>
<pre><code>$ ps f PID TTY STAT TIME COMMAND 12472 pts/0 Ss 0:00 -bash 12684 pts/0 R+ 0:00 \_ ps f </code></pre>
<p>or <code>pstree</code></p>
<pre><code>$ pstree <span>-a</span> init &#9500;&#9472;atd &#9500;&#9472;cron &#9500;&#9472;sshd -D &#9474; &#9492;&#9472;sshd &#9474; &#9492;&#9472;sshd &#9474; &#9492;&#9472;bash &#9474; &#9492;&#9472;pstree <span>-a</span> ... </code></pre>
<p>If you have ever wondered why you often see <code>bash</code> or <code>sshd</code> as parents of some of your processes, here's why.</p>
<p>This is what happens when you run, say, <code>date</code> from your <code>bash</code> shell:</p>
<ul>
<li>
<code>bash</code> creates a new process that is a copy of itself (using a <code>fork</code> system call)</li>
<li>it will then load the program from the executable file <code>/bin/date</code> into memory (using an <code>exec</code> system call)</li>
<li>
<code>bash</code> as the parent process will wait for its child to exit</li>
</ul>
<p>So the <code>/sbin/init</code> with an ID of 1 was started at boot, which spawned the SSH daemon <code>sshd</code>. When you connect to the computer, <code>sshd</code> will spawn a process for the session which in turn will launch the <code>bash</code> shell.</p>
<p>I like to use this tree view in <code>htop</code> when I'm also interested in seeing all threads.</p>
<p>Process user</p>
<p>Each process is owned by a user. Users are represented with a numeric ID.</p>
<pre><code>$ sleep 1000 &amp; [1] 2045 $ grep Uid /proc/2045/status Uid: 1000 1000 1000 1000 </code></pre>
<p>You can use the <code>id</code> command to find out the name for this user.</p>
<pre><code>$ id 1000 uid=1000(ubuntu) gid=1000(ubuntu) groups=1000(ubuntu),4(adm) </code></pre>
<p>It turns out that <code>id</code> gets this information from the <code>/etc/passwd</code> and <code>/etc/group</code> files.</p>
<pre><code>$ strace <span>-e</span> open id 1000 ... open(<span>"/etc/nsswitch.conf"</span>, O_RDONLY|O_CLOEXEC) = 3 open(<span>"/lib/x86_64-linux-gnu/libnss_compat.so.2"</span>, O_RDONLY|O_CLOEXEC) = 3 open(<span>"/lib/x86_64-linux-gnu/libnss_files.so.2"</span>, O_RDONLY|O_CLOEXEC) = 3 open(<span>"/etc/passwd"</span>, O_RDONLY|O_CLOEXEC) = 3 open(<span>"/etc/group"</span>, O_RDONLY|O_CLOEXEC) = 3 ... </code></pre>
<p>That's because the Name Service Switch (NSS) configuration file <code>/etc/nsswitch.conf</code> says to use these files to resolve names.</p>
<pre><code>$ head -n 9 /etc/nsswitch.conf <span># ...</span> passwd: compat group: compat shadow: compat </code></pre>
<p>The value of <code>compat</code> (Compatibility mode) is the same as <code>files</code> except other special entries are permitted. <code>files</code> means that the database is stored in a file (loaded by <code>libnss_files.so</code>). But you could also store your users in other databases and services or use Lightweight Directory Access Protocol (LDAP), for example.</p>
<p><code>/etc/passwd</code> and <code>/etc/group</code> are plain text files that map numeric IDs to human readable names.</p>
<pre><code>$ cat /etc/passwd root:x:0:0:root:/root:/bin/bash daemon:x:1:1:daemon:/usr/sbin:/usr/sbin/nologin ubuntu:x:1000:1000:Ubuntu:/home/ubuntu:/bin/bash $ cat /etc/group root:x:0: adm:x:4:syslog,ubuntu ubuntu:x:1000: </code></pre>
<p><code>passwd</code>? But where are the passwords?</p>
<p>They are actually in <code>/etc/shadow</code>.</p>
<pre><code>$ sudo cat /etc/shadow root:<span>$6</span><span>$mS9o0QBw</span><span>$P1ojPSTexV2PQ</span>.Z./rqzYex.k7TJE2nVeIVL0dql/:17126:0:99999:7::: daemon:*:17109:0:99999:7::: ubuntu:<span>$6</span><span>$GIfdqlb</span>/<span>$ms9ZoxfrUq455K6UbmHyOfz7DVf7TWaveyHcp</span>.:17126:0:99999:7::: </code></pre>
<p>What's that gibberish?</p>
<ul>
<li>
<code>$6$</code> is the password hashing algorithm used, in this case it stands for <code>sha512</code>
</li>
<li>followed by randomly generated salt to safeguard against rainbow table attacks</li>
<li>and finally the hash of your password + salt</li>
</ul>
<p>When you run a program, it will be run as your user. Even if the executable file is not owned by you.</p>
<p>If you'd like to run a program as <code>root</code> or another user, that's what <code>sudo</code> is for.</p>
<pre><code>$ id uid=1000(ubuntu) gid=1000(ubuntu) groups=1000(ubuntu),4(adm) $ sudo id uid=0(root) gid=0(root) groups=0(root) $ sudo -u ubuntu id uid=1000(ubuntu) gid=1000(ubuntu) groups=1000(ubuntu),4(adm) $ sudo -u daemon id uid=1(daemon) gid=1(daemon) groups=1(daemon) </code></pre>
<p>But what if you want to log in as another user to launch various commands? Use <code>sudo bash</code> or <code>sudo -u user bash</code>. You'll be able to use the shell as that user.</p>
<p>If you don't like being asked for the root password all the time, you can simply disable it by adding your user to the <code>/etc/sudoers</code> file.</p>
<p>Let's try it:</p>
<pre><code>$ <span>echo</span> <span>"<span>$USER</span> ALL=(ALL) NOPASSWD: ALL"</span> &gt;&gt; /etc/sudoers -bash: /etc/sudoers: Permission denied </code></pre>
<p>Right, only root can do it.</p>
<pre><code>$ sudo <span>echo</span> <span>"<span>$USER</span> ALL=(ALL) NOPASSWD: ALL"</span> &gt;&gt; /etc/sudoers -bash: /etc/sudoers: Permission denied </code></pre>
<p>WTF?</p>
<p>What happens here is that you are executing the <code>echo</code> command as root but appending the line to the <code>/etc/sudoers</code> file still as your user.</p>
<p>There are usually two ways around it:</p>
<ul>
<li><code>echo "$USER ALL=(ALL) NOPASSWD: ALL" | sudo tee -a /etc/sudoers</code></li>
<li><code>sudo bash -c "echo '$USER ALL=(ALL) NOPASSWD: ALL' &gt;&gt; /etc/sudoers"</code></li>
</ul>
<p>In the first case, <code>tee -a</code> will append its standard input to the file and we execute this command as root.</p>
<p>In the second case, we run bash as root and ask it to execute a command (<code>-c</code>) and the entire command will be executed as root. Note the tricky <code>"</code>/<code>'</code> business here which will dictate when the <code>$USER</code> variable will be expanded.</p>
<p>If you take a look at the <code>/etc/sudoers</code> file you will see that it begins with</p>
<pre><code>$ sudo head -n 3 /etc/sudoers <span>#</span> <span># This file MUST be edited with the 'visudo' command as root.</span> <span>#</span> </code></pre>
<p>Uh oh.</p>
<p>It's a helpful warning that says you should edit this file with <code>sudo visudo</code>. It will validate the contents of the file before saving and prevent you from making mistakes. If you don't use <code>visudo</code> and make a mistake, it may lock you out from <code>sudo</code>. Which means that you won't be able to correct your mistake!</p>
<p>Let's say you want to change your password. You can do it with the <code>passwd</code> command. It will, as we saw earlier, save the password to the <code>/etc/shadow</code> file.</p>
<p>This file is sensitive and only writeable by root:</p>
<pre><code>$ ls <span>-l</span> /etc/shadow -rw-r----- 1 root shadow 1122 Nov 27 18:52 /etc/shadow </code></pre>
<p>So how is it possible that the <code>passwd</code> program which is executed by a regular user can write to a protected file?</p>
<p>I said earlier that when you launch a process, it is owned by you, even if the owner of the executable file is another user.</p>
<p>It turns out that you can change that behavior by changing file permissions. Let's take a look.</p>
<pre><code>$ ls <span>-l</span> /usr/bin/passwd -rwsr-xr-x 1 root root 54256 Mar 29 2016 /usr/bin/passwd </code></pre>
<p>Notice the <code>s</code> letter. It was accomplished with <code>sudo chmod u+s /usr/bin/passwd</code>. It means that an executable will be launched as the the owner of the file which is root in this case.</p>
<p>You can find the so called <code>setuid</code> executables with <code>find /bin -user root -perm -u+s</code>.</p>
<p>Note that you can also do the same with group (<code>g+s</code>).</p>
<p>Process state</p>
<p>We are next going to look at the process state column in <code>htop</code> which is denoted simply with the letter <code>S</code>.</p>
<p>Here are the possible values:</p>
<pre><code>R running or runnable (on run queue) S interruptible sleep (waiting for an event to complete) D uninterruptible sleep (usually IO) Z defunct ("zombie") process, terminated but not reaped by its parent T stopped by job control signal t stopped by debugger during the tracing X dead (should never be seen) </code></pre>
<p>I've ordered them by how often I see them.</p>
<p>Note that when you run <code>ps</code>, it will also show substates like <code>Ss</code>, <code>R+</code>, <code>Ss+</code>, etc.</p>
<pre><code>$ ps x PID TTY STAT TIME COMMAND 1688 ? Ss 0:00 /lib/systemd/systemd --user 1689 ? S 0:00 (sd-pam) 1724 ? S 0:01 sshd: vagrant@pts/0 1725 pts/0 Ss 0:00 -bash 2628 pts/0 R+ 0:00 ps x </code></pre>
<p>R - running or runnable (on run queue)</p>
<p>In this state, the process is currently running or on a run queue waiting to run.</p>
<p>What does it mean to run?</p>
<p>When you compile the source code of a program that you've written, that machine code is CPU instructions. It is saved to a file that can be executed. When you launch a program, it is loaded into memory and then the CPU executes these instructions.</p>
<p>Basically it means that the CPU is physically executing instructions. Or, in other words, crunching numbers.</p>
<p>S - interruptible sleep (waiting for an event to complete)</p>
<p>This means that the code instructions of this process are not being executed on the CPU. Instead, this process is waiting for something - an event or a condition - to happen. When an event happens, the kernel sets the state to running.</p>
<p>One example is the <code>sleep</code> utily from coreutils. It will sleep for a specific number of seconds (approximately).</p>
<pre><code>$ sleep 1000 &amp; [1] 10089 $ ps f PID TTY STAT TIME COMMAND 3514 pts/1 Ss 0:00 -bash 10089 pts/1 S 0:00 \_ sleep 1000 10094 pts/1 R+ 0:00 \_ ps f </code></pre>
<p>So this is <em>interruptible</em> sleep. How can we interrupt it?</p>
<p>By sending a signal.</p>
<p>You can send a signal in <code>htop</code> by hitting <code>F9</code> and then choosing one of the signals in the menu on the left.</p>
<p>Sending a signal is also known as <code>kill</code>. That's because <code>kill</code> is a system call that can send a signal to a process. There is a program <code>/bin/kill</code> that can make this system call from userland and the default signal to use is <code>TERM</code> which will ask the process to terminate or in other words try to kill it.</p>
<p>Signal is just a number. Numbers are hard to remember so we give them names. Signal names are usually written in uppercase and may be prefixed with <code>SIG</code>.</p>
<p>Some commonly used signals are <code>INT</code>, <code>KILL</code>, <code>STOP</code>, <code>CONT</code>, <code>HUP</code>.</p>
<p>Let's interrupt the sleep process by sending the <code>INT</code> aka <code>SIGINT</code> aka <code>2</code> aka <code>Terminal interrupt</code> signal.</p>
<pre><code>$ <span>kill</span> -INT 10089 [1]+ Interrupt sleep 1000 </code></pre>
<p>This is also what happens When you hit <code>CTRL</code>+<code>C</code> on your keyboard. <code>bash</code> will the send the foreground process the <code>SIGINT</code> signal just like we just did manually.</p>
<p>By the way, in <code>bash</code>, <code>kill</code> is a built-in command, even though there is <code>/bin/kill</code> on most systems. Why? It allows processes to be killed if the limit on processes that you can create is reached.</p>
<p>These commands do the same thing:</p>
<ul>
<li><code>kill -INT 10089</code></li>
<li><code>kill -2 10089</code></li>
<li><code>/bin/kill -2 10089</code></li>
</ul>
<p>Another useful signal to know is <code>SIGKILL</code> aka <code>9</code>. You may have used it to kill a process that didn't respond to your frantic <code>CTRL</code>+<code>C</code> keyboard presses.</p>
<p>When you write a program, you can set up signal handlers that are functions that will be called when your process receives a signal. In other words, you can catch the signal and then do something, for example, clean up and shut down gracefully. So sending <code>SIGINT</code> (the user wants to interrupt a process) and <code>SIGTERM</code> (the user wants to terminate the process) does not mean that the process will be terminated.</p>
<p>You may have seen this exception when running Python scripts:</p>
<pre><code>$ python -c <span>'import sys; sys.stdin.read()'</span> ^C Traceback (most recent call last): File <span>"&lt;string&gt;"</span>, line 1, <span>in</span> &lt;module&gt; KeyboardInterrupt </code></pre>
<p>You can tell the kernel to forcefully terminate a process and not give it a change to respond by sending the <code>KILL</code> signal:</p>
<pre><code>$ sleep 1000 &amp; [1] 2658 $ <span>kill</span> -9 2658 [1]+ Killed sleep 1000 </code></pre>
<p>D - uninterruptible sleep (usually IO)</p>
<p>Unlike interruptible sleep, you cannot wake up this process with a signal. That is why many people dread seeing this state. You can't kill such processes because killing means sending <code>SIGKILL</code> signals to processes.</p>
<p>This state is used if the process must wait without interruption or when the event is expected to occur quickly. Like reading to/from a disk. But that should only happen for a fraction of a second.</p>
<p>Here is a <a href="http://stackoverflow.com/questions/223644/what-is-an-uninterruptable-process">nice answer on StackOverflow</a>.</p>
<blockquote>
<p>Uninterruptable processes are USUALLY waiting for I/O following a page fault. The process/task cannot be interrupted in this state, because it can't handle any signals; if it did, another page fault would happen and it would be back where it was.</p>
</blockquote>
<p>In other words, this could happen if you are using Network File System (NFS) and it takes a while to read and write from it.</p>
<p>Or in my experience it can also mean that some of the processes are swapping a lot which means you have too little available memory.</p>
<p>Let's try to get a process to go into uninterruptible sleep.</p>
<p><code>8.8.8.8</code> is a public DNS server provided by Google. They do not have an open NFS on there. But that won't stop us.</p>
<pre><code>$ sudo mount 8.8.8.8:/tmp /tmp &amp; [1] 12646 $ sudo ps x | grep mount.nfs 12648 pts/1 D 0:00 /sbin/mount.nfs 8.8.8.8:/tmp /tmp -o rw </code></pre>
<p>How to find out what's causing this? <code>strace</code>!</p>
<p>Let's <code>strace</code> the command in the output of <code>ps</code> above.</p>
<pre><code>$ sudo strace /sbin/mount.nfs 8.8.8.8:/tmp /tmp -o rw ... mount(<span>"8.8.8.8:/tmp"</span>, <span>"/tmp"</span>, <span>"nfs"</span>, 0, ... </code></pre>
<p>So the <code>mount</code> system call is blocking the process.</p>
<p>If you're wondering, you can run <code>mount</code> with an <code>intr</code> option to run as interruptible: <code>sudo mount 8.8.8.8:/tmp /tmp -o intr</code>.</p>
<p>Z - defunct ("zombie") process, terminated but not reaped by its parent</p>
<p>When a process ends via <code>exit</code> and it still has child processes, the child processes become zombie processes.</p>
<ul>
<li>If zombie processes exist for a short time, it is perfectly normal</li>
<li>Zombie processes that exist for a long time may indicate a bug in a program</li>
<li>Zombie processes don't consume memory, just a process ID</li>
<li>You can't <code>kill</code> a zombie process</li>
<li>You can ask nicely the parent process to reap the zombies (the <code>SIGCHLD</code> signal)</li>
<li>You can <code>kill</code> the zombie's parent process to get rid of the parent and its zombies</li>
</ul>
<p>I am going to write some C code to show this.</p>
<p>Here is our program.</p>
<pre><code><span>#<span>include</span> <span>&lt;stdio.h&gt;</span></span> <span>#<span>include</span> <span>&lt;stdlib.h&gt;</span></span> <span>#<span>include</span> <span>&lt;unistd.h&gt;</span></span> <span><span>int</span> <span>main</span><span>()</span> </span>{ <span>printf</span>(<span>"Running\n"</span>); <span>int</span> pid = fork(); <span>if</span> (pid == <span>0</span>) { <span>printf</span>(<span>"I am the child process\n"</span>); <span>printf</span>(<span>"The child process is exiting now\n"</span>); <span>exit</span>(<span>0</span>); } <span>else</span> { <span>printf</span>(<span>"I am the parent process\n"</span>); <span>printf</span>(<span>"The parent process is sleeping now\n"</span>); sleep(<span>20</span>); <span>printf</span>(<span>"The parent process is finished\n"</span>); } <span>return</span> <span>0</span>; } </code></pre>
<p>Let's install the GNU C Compiler (GCC).</p>
<pre><code>sudo apt install -y gcc </code></pre>
<p>Compile it and then run it</p>
<pre><code>gcc zombie.c -o zombie ./zombie </code></pre>
<p>Look at the process tree</p>
<pre><code>$ ps f PID TTY STAT TIME COMMAND 3514 pts/1 Ss 0:00 -bash 7911 pts/1 S+ 0:00 \_ ./zombie 7912 pts/1 Z+ 0:00 \_ [zombie] &lt;defunct&gt; 1317 pts/0 Ss 0:00 -bash 7913 pts/0 R+ 0:00 \_ ps f </code></pre>
<p>We got our zombie!</p>
<p>When the parent process is done, the zombie is gone.</p>
<pre><code>$ ps f PID TTY STAT TIME COMMAND 3514 pts/1 Ss+ 0:00 -bash 1317 pts/0 Ss 0:00 -bash 7914 pts/0 R+ 0:00 \_ ps f </code></pre>
<p>If you replaced <code>sleep(20)</code> with <code>while (true) ;</code> then the zombie would be gone right away.</p>
<p>With <code>exit</code>, all of the memory and resources associated with it are deallocated so they can be used by other processes.</p>
<p>Why keep the zombie processes around then?</p>
<p>The parent process has the option to find out its child process exit code (in a signal handler) with the <code>wait</code> system call. If a process is sleeping, then it needs to wait for it to wake up.</p>
<p>Why not simply forcefully wake it up and kill it? For the same reason, you don't toss your child in the trash when you're tired of it. Bad things could happen.</p>
<p>T - stopped by job control signal</p>
<p>I have opened two terminal windows and I can look at my user's processes with <code>ps u</code>.</p>
<pre><code>$ ps u USER PID %CPU %MEM VSZ RSS TTY STAT START TIME COMMAND ubuntu 1317 0.0 0.9 21420 4992 pts/0 Ss+ Jun07 0:00 -bash ubuntu 3514 1.5 1.0 21420 5196 pts/1 Ss 07:28 0:00 -bash ubuntu 3528 0.0 0.6 36084 3316 pts/1 R+ 07:28 0:00 ps u </code></pre>
<p>I will omit the <code>-bash</code> and <code>ps u</code> processes from the output below.</p>
<p>Now run <code>cat /dev/urandom &gt; /dev/null</code> in one terminal window. Its state is <code>R+</code> which means that it is running.</p>
<pre><code>$ ps u USER PID %CPU %MEM VSZ RSS TTY STAT START TIME COMMAND ubuntu 3540 103 0.1 6168 688 pts/1 R+ 07:29 0:04 cat /dev/urandom </code></pre>
<p>Press <code>CTRL</code>+<code>Z</code> to stop the process.</p>
<pre><code>$ <span># CTRL+Z</span> [1]+ Stopped cat /dev/urandom &gt; /dev/null $ ps aux USER PID %CPU %MEM VSZ RSS TTY STAT START TIME COMMAND ubuntu 3540 86.8 0.1 6168 688 pts/1 T 07:29 0:15 cat /dev/urandom </code></pre>
<p>Its state is now <code>T</code>.</p>
<p>Run <code>fg</code> in the first terminal to resume it.</p>
<p>Another way to stop a process like this is to send the <code>STOP</code> signal with <code>kill</code> to the process. To resume the execution of the process, you can use the <code>CONT</code> signal.</p>
<p>t - stopped by debugger during the tracing</p>
<p>First, install the GNU Debugger (gdb)</p>
<pre><code>sudo apt install -y gdb </code></pre>
<p>Run a program that will listen for incoming network connections on port 1234.</p>
<pre><code>$ nc <span>-l</span> 1234 &amp; [1] 3905 </code></pre>
<p>It is sleeping meaning it is waiting for data from the network.</p>
<pre><code>$ ps u USER PID %CPU %MEM VSZ RSS TTY STAT START TIME COMMAND ubuntu 3905 0.0 0.1 9184 896 pts/0 S 07:41 0:00 nc <span>-l</span> 1234 </code></pre>
<p>Run the debugger and attach it to the process with ID <code>3905</code>.</p>
<pre><code>sudo gdb -p 3905 </code></pre>
<p>You will see that the state is <code>t</code> which means that this process is being traced in the debugger.</p>
<pre><code>$ ps u USER PID %CPU %MEM VSZ RSS TTY STAT START TIME COMMAND ubuntu 3905 0.0 0.1 9184 896 pts/0 t 07:41 0:00 nc <span>-l</span> 1234 </code></pre>
<p>Process time</p>
<p>Linux is a multitasking operating system which means that even when you have a single CPU, you can run several processes at the same time. You can connect to your server via SSH and look at the output of <code>htop</code> while your web server is delivering the content of your blog to your readers over the internet.</p>
<p>How is that possible when a single CPU can only execute one instruction at a time?</p>
<p>The answer is time sharing.</p>
<p>One process runs for a bit of time, then it is suspended while the other processes waiting to run take turns running for a while. The bit of time a process runs is called the time slice.</p>
<p>The time slice is usually a few milliseconds so you don't really notice it that much when your system is not under high load. (It'd be really interesting to find out how long time slices usually are in Linux.)</p>
<p>This should help explain why the load average is the average number of running processes. If you have just one core and the load average is <code>1.0</code>, the CPU has been utilized at 100%. If the load average is higher than <code>1.0</code>, it means that the number of processes wanting to run is higher than the CPU can run so you may experience slow downs or delays. If the load is lower than <code>1.0</code>, it means the CPU is sometimes idleing and not doing anything.</p>
<p>This should also give you a clue why sometimes the running time of a process that's been running for 10 seconds is higher or lower than exactly 10 seconds.</p>
<p>Process niceness and priority</p>
<p>When you have more tasks to run than the number of available CPU cores, you somehow have to decide which tasks to run next and which ones to keep waiting. This is what the task scheduler is responsible for.</p>
<p>The scheduler in the Linux kernel is reponsible for choosing which process on a run queue to pick next and it depends on the scheduler algorithm used in the kernel.</p>
<p>You can't generally influence the scheduler but you can let it know which processes are more important to you and the scheduler may take it into account.</p>
<p>Niceness (<code>NI</code>) is user-space priority to processes, ranging from -20 which is the highest priority to 19 which is the lowest priority. It can be confusing but you can think that a nice process yields to a less nice process. So the nicer a process is, the more it yields.</p>
<p>From what I've pieced together by reading StackOverflow and other sites, a niceness level increase by 1 should yield a 10% more CPU time to the process.</p>
<p>The priority (<code>PRI</code>) is the kernel-space priority that the Linux kernel is using. Priorities range from 0 to 139 and the range from 0 to 99 is real time and 100 to 139 for users.</p>
<p>You can change the nicesness and the kernel takes it into account but you cannot change the priority.</p>
<p>The relation between the nice value and priority is:</p>
<pre><code>PR = 20 + NI </code></pre>
<p>so the value of <code>PR = 20 + (-20 to +19)</code> is 0 to 39 that maps 100 to 139.</p>
<p>You can set the niceness of a process before launching it.</p>
<pre><code>nice -n niceness program </code></pre>
<p>Change the nicencess when a program is already running with <code>renice</code>.</p>
<pre><code>renice -n niceness -p PID </code></pre>
<p>Here is what the CPU usage colors mean:</p>
<ul>
<li>Blue: Low priority threads (nice &gt; 0)</li>
<li>Green: Normal priority threads</li>
<li>Red: Kernel threads</li>
</ul>
<p><a href="http://askubuntu.com/questions/656771/process-niceness-vs-priority">http://askubuntu.com/questions/656771/process-niceness-vs-priority</a></p>
<p>Memory usage - VIRT/RES/SHR/MEM</p>
<p>A process has the illusion of being the only one in memory. This is accomplished by using virtual memory.</p>
<p>A process does not have direct access to the physical memory. Instead, it has its own virtual address space and the kernel translates the virtual memory addresses to physical memory or can map some of it to disk. This is why it can look like processes use more memory than you have installed on your computer.</p>
<p>The point I want to make here is that it is not very straightforward to figure out how much memory a process takes up. Do you also want to count the shared libraries or disk mapped memory? But the kernel provides and <code>htop</code> shows some information that can help you estimate memory usage.</p>
<p>Here is what the memory usage colors mean:</p>
<ul>
<li>Green: Used memory</li>
<li>Blue: Buffers</li>
<li>Orange: Cache</li>
</ul>
<p>VIRT/VSZ - Virtual Image</p>
<blockquote>
<p>The total amount of virtual memory used by the task. It includes all code, data and shared libraries plus pages that have been swapped out and pages that have been mapped but not used.</p>
</blockquote>
<p><code>VIRT</code> is virtual memory usage. It includes everything, including memory mapped files.</p>
<p>If an application requests 1 GB of memory but uses only 1 MB, then <code>VIRT</code> will report 1 GB. If it <code>mmap</code>s a 1 GB file and never uses it, <code>VIRT</code> will also report 1 GB.</p>
<p>Most of the time, this is not a useful number.</p>
<p>RES/RSS - Resident size</p>
<blockquote>
<p>The non-swapped physical memory a task has used.</p>
</blockquote>
<p><code>RES</code> is resident memory usage i.e. what's currently in the physical memory.</p>
<p>While <code>RES</code> can be a better indicator of how much memory a process is using than <code>VIRT</code>, keep in mind that</p>
<ul>
<li>this does not include the swapped out memory</li>
<li>some of the memory may be shared with other processes</li>
</ul>
<p>If a process uses 1 GB of memory and it calls <code>fork()</code>, the result of forking will be two processes whose <code>RES</code> is both 1 GB but only 1 GB will actually be used since Linux uses copy-on-write.</p>
<p>SHR - Shared Mem size</p>
<blockquote>
<p>The amount of shared memory used by a task.<br>
It simply reflects memory that could be potentially shared with other processes.</p>
</blockquote>
<pre><code><span>#<span>include</span> <span>&lt;stdio.h&gt;</span></span> <span>#<span>include</span> <span>&lt;stdlib.h&gt;</span></span> <span>#<span>include</span> <span>&lt;unistd.h&gt;</span></span> <span><span>int</span> <span>main</span><span>()</span> </span>{ <span>printf</span>(<span>"Started\n"</span>); sleep(<span>10</span>); <span>size_t</span> memory = <span>10</span> * <span>1024</span> * <span>1024</span>; <span>// 10 MB</span> <span>char</span>* buffer = <span>malloc</span>(memory); <span>printf</span>(<span>"Allocated 10M\n"</span>); sleep(<span>10</span>); <span>for</span> (<span>size_t</span> i = <span>0</span>; i &lt; memory/<span>2</span>; i++) buffer[i] = <span>42</span>; <span>printf</span>(<span>"Used 5M\n"</span>); sleep(<span>10</span>); <span>int</span> pid = fork(); <span>printf</span>(<span>"Forked\n"</span>); sleep(<span>10</span>); <span>if</span> (pid != <span>0</span>) { <span>for</span> (<span>size_t</span> i = memory/<span>2</span>; i &lt; memory/<span>2</span> + memory/<span>5</span>; i++) buffer[i] = <span>42</span>; <span>printf</span>(<span>"Child used extra 2M\n"</span>); } sleep(<span>10</span>); <span>return</span> <span>0</span>; } </code></pre>
<pre><code>fallocate <span>-l</span> 10G gcc -std=c99 mem.c -o mem ./mem </code></pre>
<pre><code>Process Message VIRT RES SHR main Started 4200 680 604 main Allocated 10M 14444 680 604 main Used 5M 14444 6168 1116 main Forked 14444 6168 1116 child Forked 14444 5216 0 main Child used extra 2M 8252 1116 child Child used extra 2M 5216 0 </code></pre>
<p>TODO: I should finish this.</p>
<p>MEM% - Memory usage</p>
<blockquote>
<p>A task's currently used share of available physical memory.</p>
</blockquote>
<p>This is <code>RES</code> divided by the total RAM you have.</p>
<p>If <code>RES</code> is <code>400M</code> and you have 8 gigabytes of RAM, <code>MEM%</code> will be <code>400/8192*100</code> = <code>4.88%</code>.</p>
<p>Processes</p>
<p>I launched a Digital Ocean droplet with Ubuntu Server.</p>
<p>What are the processes that are started at boot?</p>
<p>Do you actually need them?</p>
<p>Here are my research notes on the processes that are run at startup on a fresh Digital Ocean droplet with Ubuntu Server 16.04.1 LTS x64.</p>
<p>Before</p>
<p><img src="https://peteris.rocks/blog/htop/canyoukillit-before.png"></p>
<p><code>/sbin/init</code></p>
<blockquote>
<p>The /sbin/init program (also called init) coordinates the rest of the boot process and configures the environment for the user.</p>
<p>When the init command starts, it becomes the parent or grandparent of all of the processes that start up automatically on the system.</p>
</blockquote>
<p>Is it systemd?</p>
<pre><code>$ dpkg -S /sbin/init systemd-sysv: /sbin/init </code></pre>
<p>Yes, it is.</p>
<p>What happens if you kill it?</p>
<p>Nothing.</p>
<p><code>/lib/systemd/systemd-journald</code></p>
<blockquote>
<p>systemd-journald is a system service that collects and stores logging data. It creates and maintains structured, indexed journals based on logging information that is received from a variety of sources.</p>
</blockquote>
<p>In other words:</p>
<blockquote>
<p>One of the main changes in journald was to replace simple plain text log files with a special file format optimized for log messages. This file format allows system administrators to access relevant messages more efficiently. It also brings some of the power of database-driven centralized logging implementations to individual systems.</p>
</blockquote>
<p>You are supposed to use the <code>journalctl</code> command to query log files.</p>
<ul>
<li>
<code>journalctl _COMM=sshd</code> logs by sshd</li>
<li>
<code>journalctl _COMM=sshd -o json-pretty</code> logs by sshd in JSON</li>
<li><code>journalctl --since "2015-01-10" --until "2015-01-11 03:00"</code></li>
<li><code>journalctl --since 09:00 --until "1 hour ago"</code></li>
<li><code>journalctl --since yesterday</code></li>
<li>
<code>journalctl -b</code> logs since boot</li>
<li>
<code>journalctl -f</code> to follow logs</li>
<li><code>journalctl --disk-usage</code></li>
<li><code>journalctl --vacuum-size=1G</code></li>
</ul>
<p>Pretty cool.</p>
<p>It looks like it is not possible to remove or disable this service, you can only turn off logging.</p>
<blockquote>
<p>The lvmetad daemon caches LVM metadata, so that LVM commands can read metadata without scanning disks.</p>
<p>Metadata caching can be an advantage because scanning disks is time consuming and may interfere with the normal work of the system and disks.</p>
</blockquote>
<p>But what is LVM (Logical Volume Management)?</p>
<blockquote>
<p>You can think of LVM as "dynamic partitions", meaning that you can create/resize/delete LVM "partitions" (they're called "Logical Volumes" in LVM-speak) from the command line while your Linux system is running: no need to reboot the system to make the kernel aware of the newly-created or resized partitions.</p>
</blockquote>
<p>It sounds like you should keep it if you are using LVM.</p>
<pre><code>$ lvscan $ sudo apt remove lvm2 -y --purge </code></pre>
<p><code>/lib/systemd/udevd</code></p>
<blockquote>
<p>systemd-udevd listens to kernel uevents. For every event, systemd-udevd executes matching instructions specified in udev rules.</p>
<p>udev is a device manager for the Linux kernel. As the successor of devfsd and hotplug, udev primarily manages device nodes in the /dev directory.</p>
</blockquote>
<p>So this service manages <code>/dev</code>.</p>
<p>I am not sure if I need it running on a virtual server.</p>
<p><code>/lib/systemd/timesyncd</code></p>
<blockquote>
<p>systemd-timesyncd is a system service that may be used to synchronize the local system clock with a remote Network Time Protocol server.</p>
</blockquote>
<p>So this replaces <code>ntpd</code>.</p>
<pre><code>$ timedatectl status Local time: Fri 2016-08-26 11:38:21 UTC Universal time: Fri 2016-08-26 11:38:21 UTC RTC time: Fri 2016-08-26 11:38:20 Time zone: Etc/UTC (UTC, +0000) Network time on: yes NTP synchronized: yes RTC <span>in</span> <span>local</span> TZ: no </code></pre>
<p>If we take a look at the open ports on this server:</p>
<pre><code>$ sudo netstat -nlput Active Internet connections (only servers) Proto Recv-Q Send-Q Local Address Foreign Address State PID/Program name tcp 0 0 0.0.0.0:22 0.0.0.0:* LISTEN 2178/sshd tcp6 0 0 :::22 :::* LISTEN 2178/sshd </code></pre>
<p>Lovely!</p>
<p>Previously on Ubuntu 14.04 it was</p>
<pre><code>$ sudo apt-get install ntp -y $ sudo netstat -nlput Active Internet connections (only servers) Proto Recv-Q Send-Q Local Address Foreign Address State PID/Program name tcp 0 0 0.0.0.0:22 0.0.0.0:* LISTEN 1380/sshd tcp6 0 0 :::22 :::* LISTEN 1380/sshd udp 0 0 10.19.0.6:123 0.0.0.0:* 2377/ntpd udp 0 0 139.59.256.256:123 0.0.0.0:* 2377/ntpd udp 0 0 127.0.0.1:123 0.0.0.0:* 2377/ntpd udp 0 0 0.0.0.0:123 0.0.0.0:* 2377/ntpd udp6 0 0 fe80::601:6aff:fxxx:123 :::* 2377/ntpd udp6 0 0 ::1:123 :::* 2377/ntpd udp6 0 0 :::123 :::* 2377/ntpd </code></pre>
<p>Ugh.</p>
<p><code>/usr/sbin/atd -f</code></p>
<blockquote>
<p>atd - run jobs queued for later execution. atd runs jobs queued by at.</p>
<p>at and batch read commands from standard input or a specified file which are to be executed at a later time</p>
</blockquote>
<p>Unlike cron, which schedules jobs that are repeated periodically, <code>at</code> runs a job at a specific time once.</p>
<pre><code>$ <span>echo</span> <span>"touch /tmp/yolo.txt"</span> | at now + 1 minute job 1 at Fri Aug 26 10:44:00 2016 $ atq 1 Fri Aug 26 10:44:00 2016 a root $ sleep 60 &amp;&amp; ls /tmp/yolo.txt /tmp/yolo.txt </code></pre>
<p>I've actually never used it until now.</p>
<pre><code>sudo apt remove at -y --purge </code></pre>
<p><code>/usr/lib/snapd/snapd</code></p>
<blockquote>
<p>Snappy Ubuntu Core is a new rendition of Ubuntu with transactional updates - a minimal server image with the same libraries as today&rsquo;s Ubuntu, but applications are provided through a simpler mechanism.</p>
</blockquote>
<p>What?</p>
<blockquote>
<p>Developers from multiple Linux distributions and companies today announced collaboration on the &ldquo;snap&rdquo; universal Linux package format, enabling a single binary package to work perfectly and securely on any Linux desktop, server, cloud or device.</p>
</blockquote>
<p>Apparently it is a simplified deb package and you're supposted to bundle all dependencies in a single snap that you can distribute.</p>
<p>I've never used snappy to deploy or distribute applications on servers.</p>
<pre><code>sudo apt remove snapd -y --purge </code></pre>
<p><code>/usr/bin/dbus-daemon</code></p>
<blockquote>
<p>In computing, D-Bus or DBus is an inter-process communication (IPC) and remote procedure call (RPC) mechanism that allows communication between multiple computer programs (that is, processes) concurrently running on the same machine</p>
</blockquote>
<p>My understanding is that you need it for desktop environments but on a server to run web apps?</p>
<pre><code>sudo apt remove dbus -y --purge </code></pre>
<p>I wonder what time it is and whether it is being synchronized with NTP?</p>
<pre><code>$ timedatectl status Failed to create bus connection: No such file or directory </code></pre>
<p>Oops. Should probably keep this.</p>
<p><code>/lib/systemd/systemd-logind</code></p>
<blockquote>
<p>systemd-logind is a system service that manages user logins.</p>
</blockquote>
<p><code>/usr/sbin/cron -f</code></p>
<blockquote>
<p>cron - daemon to execute scheduled commands (Vixie Cron)</p>
<p><code>-f</code> Stay in foreground mode, don't daemonize.</p>
</blockquote>
<p>You can schedule tasks to run periodically with cron.</p>
<p>Use <code>crontab -e</code> to edit the configuration for your user or on Ubuntu I tend to use the <code>/etc/cron.hourly</code>, <code>/etc/cron.daily</code>, etc. directories.</p>
<p>You can see the log files with</p>
<ul>
<li>
<code>grep cron /var/log/syslog</code> or</li>
<li>
<code>journalctl _COMM=cron</code> or even</li>
<li><code>journalctl _COMM=cron --since="date" --until="date"</code></li>
</ul>
<p>You'll probably want to keep cron.</p>
<p>But if you don't, then you should stop and disable the service:</p>
<pre><code>sudo systemctl stop cron sudo systemctl <span>disable</span> cron </code></pre>
<p>Because otherwise when trying to remove it with <code>apt remove cron</code> it will try to install postfix!</p>
<pre><code>$ sudo apt remove cron The following packages will be REMOVED: cron The following NEW packages will be installed: anacron bcron bcron-run fgetty libbg1 libbg1-doc postfix runit ssl-cert ucspi-unix </code></pre>
<p>It looks like cron needs a mail transport agent (MTA) to send emails.</p>
<pre><code>$ apt show cron Package: cron Version: 3.0pl1-128ubuntu2 ... Suggests: anacron (&gt;= 2.0-1), logrotate, checksecurity, exim4 | postfix | mail-transport-agent $ apt depends cron cron ... Suggests: anacron (&gt;= 2.0-1) Suggests: logrotate Suggests: checksecurity |Suggests: exim4 |Suggests: postfix Suggests: &lt;mail-transport-agent&gt; ... exim4-daemon-heavy postfix </code></pre>
<p><code>/usr/sbin/rsyslogd -n</code></p>
<blockquote>
<p>Rsyslogd is a system utility providing support for message logging.</p>
</blockquote>
<p>In another words, it's what populates log files in <code>/var/log/</code> like <code>/var/log/auth.log</code> for authentication messages like SSH login attempts.</p>
<p>The configuration files are in <code>/etc/rsyslog.d</code>.</p>
<p>You can also configure rsyslogd to send log files to a remote server and implement centralized logging.</p>
<p>You can use the <code>logger</code> command to log messages to <code>/var/log/syslog</code> in background scripts such as those that are run at boot.</p>
<pre><code><span>#!/bin/bash </span> logger Starting doing something <span># NFS, get IPs, etc.</span> logger Done doing something </code></pre>
<p>Right, but we already have <code>systemd-journald</code> running. Do we need <code>rsyslogd</code> as well?</p>
<blockquote>
<p>Rsyslog and Journal, the two logging applications present on your system, have several distinctive features that make them suitable for specific use cases. In many situations it is useful to combine their capabilities, for example to create structured messages and store them in a file database. A communication interface needed for this cooperation is provided by input and output modules on the side of Rsyslog and by the Journal's communication socket.</p>
</blockquote>
<p>So, maybe? I am going to keep it just in case.</p>
<p><code>/usr/sbin/acpid</code></p>
<blockquote>
<p>acpid - Advanced Configuration and Power Interface event daemon</p>
<p>acpid is designed to notify user-space programs of ACPI events. acpid should be started during the system boot, and will run as a background process, by default.</p>
<p>In computing, the Advanced Configuration and Power Interface (ACPI) specification provides an open standard that operating systems can use to perform discovery and configuration of computer hardware components, to perform power management by, for example, putting unused components to sleep, and to do status monitoring.</p>
</blockquote>
<p>But I'm on a virtual server that I don't intend to suspend/resume.</p>
<p>I am going to remove it for fun and see what happens.</p>
<pre><code>sudo apt remove acpid -y --purge </code></pre>
<p>I was able to successfully <code>reboot</code> the droplet but after <code>halt</code> Digital Ocean thought it was still on so I had to Power Off using the web interface.</p>
<p>So I should probably keep this.</p>
<p><code>/usr/bin/lxcfs /var/lib/lxcfs/</code></p>
<blockquote>
<p>Lxcfs is a fuse filesystem mainly designed for use by lxc containers. On a Ubuntu 15.04 system, it will be used by default to provide two things: first, a virtualized view of some /proc files; and secondly, filtered access to the host&rsquo;s cgroup filesystems.</p>
<p>In summary, on a 15.04 host, you can now create a container the usual way, lxc-create ... The resulting container will have &ldquo;correct&rdquo; results for uptime, top, etc.</p>
<p>It&rsquo;s basically a userspace workaround to changes which were deemed unreasonable to do in the kernel. It makes containers feel much more like separate systems than they would without it.</p>
</blockquote>
<p>Not using LXC containers? You can remove it with</p>
<pre><code>sudo apt remove lxcfs -y --purge </code></pre>
<p><code>/usr/lib/accountservice/accounts-daemon</code></p>
<blockquote>
<p>The AccountsService package provides a set of D-Bus interfaces for querying and manipulating user account information and an implementation of these interfaces based on the usermod(8), useradd(8) and userdel(8) commands.</p>
</blockquote>
<p>When I removed DBus it broke <code>timedatectl</code>, I wonder what removing this service will break.</p>
<pre><code>sudo apt remove accountsservice -y --purge </code></pre>
<p>Time will tell.</p>
<p><code>/sbin/mdadm</code></p>
<blockquote>
<p>mdadm is a Linux utility used to manage and monitor software RAID devices.</p>
<p>The name is derived from the md (multiple device) device nodes it administers or manages, and it replaced a previous utility mdctl. The original name was "Mirror Disk", but was changed as the functionality increased.</p>
<p>RAID is a method of using multiple hard drives to act as one. There are two purposes of RAID: 1) Expand drive capacity: RAID 0. If you have 2 x 500 GB HDD then total space become 1 TB. 2) Prevent data loss in case of drive failure: For example RAID 1, RAID 5, RAID 6, and RAID 10.</p>
</blockquote>
<p>You can remove it with</p>
<pre><code>sudo apt remove mdadm -y --purge </code></pre>
<p><code>/usr/lib/policykit-1/polkitd --no-debug</code></p>
<blockquote>
<p>polkitd &mdash; PolicyKit daemon</p>
<p>polkit - Authorization Framework</p>
</blockquote>
<p>My understanding is that this is like fine-grained sudo. You can allow non privilegded users to do certain actions as root. For instance, reboot your computer when you're running Linux on a desktop computer.</p>
<p>But I'm running a server. You can remove it with</p>
<pre><code>sudo apt remove policykit-1 -y --purge </code></pre>
<p>Still wondering if this breaks something.</p>
<p><code>/usr/sbin/sshd -D</code></p>
<blockquote>
<p>sshd (OpenSSH Daemon) is the daemon program for ssh.</p>
<p>-D When this option is specified, sshd will not detach and does not become a daemon. This allows easy monitoring of sshd.</p>
</blockquote>
<p><code>/sbin/iscsid</code></p>
<p>iscsid is the daemon (system service) that runs in the background, acting on iSCSI configuration, and managing the connections. From its manpage:</p>
<blockquote>
<p>The iscsid implements the control path of iSCSI protocol, plus some management facilities. For example, the daemon could be configured to automatically re-start discovery at startup, based on the contents of persistent iSCSI database.</p>
</blockquote>
<p><a href="http://unix.stackexchange.com/questions/216239/iscsi-vs-iscsid-services">http://unix.stackexchange.com/questions/216239/iscsi-vs-iscsid-services</a></p>
<p>I had never heard of iSCSI:</p>
<blockquote>
<p>In computing, iSCSI (Listeni/a&#618;&#712;sk&#652;zi/ eye-skuz-ee) is an acronym for Internet Small Computer Systems Interface, an Internet Protocol (IP)-based storage networking standard for linking data storage facilities.</p>
<p>By carrying SCSI commands over IP networks, iSCSI is used to facilitate data transfers over intranets and to manage storage over long distances. iSCSI can be used to transmit data over local area networks (LANs), wide area networks (WANs), or the Internet and can enable location-independent data storage and retrieval.</p>
<p>The protocol allows clients (called initiators) to send SCSI commands (CDBs) to SCSI storage devices (targets) on remote servers. It is a storage area network (SAN) protocol, allowing organizations to consolidate storage into data center storage arrays while providing hosts (such as database and web servers) with the illusion of locally attached disks.</p>
</blockquote>
<p>You can remove it with</p>
<pre><code>sudo apt remove open-iscsi -y --purge </code></pre>
<p><code>/sbin/agetty --noclear tty1 linux</code></p>
<blockquote>
<p>agetty - alternative Linux getty</p>
<p>getty, short for "get tty", is a Unix program running on a host computer that manages physical or virtual terminals (TTYs). When it detects a connection, it prompts for a username and runs the 'login' program to authenticate the user.</p>
<p>Originally, on traditional Unix systems, getty handled connections to serial terminals (often Teletype machines) connected to a host computer. The tty part of the name stands for Teletype, but has come to mean any type of text terminal.</p>
</blockquote>
<p>This allows you to log in when you are physically at the server. In Digital Ocean, you can click on <code>Console</code> in the droplet details and you will be able to interact with this terminal in your browser (it's a VNC connection I think).</p>
<p>In the old days, you'd see a bunch of ttys started a system boot (configured in <code>/etc/inittab</code>), but nowadays they are spun up on demand by systemd.</p>
<p>For fun, I removed this configuration file that launches and generates <code>agetty</code>:</p>
<pre><code>sudo rm /etc/systemd/system/getty.target.wants/getty@tty1.service sudo rm /lib/systemd/system/getty@.service </code></pre>
<p>When I rebooted the server, I could still connect to it via SSH but I was no longer able to log in from the Digital Ocean web console.</p>
<p><img alt="Screenshot of htop" src="https://peteris.rocks/blog/htop/htop-login.png"></p>
<p><code>sshd: root@pts/0</code> &amp; <code>-bash</code> &amp; <code>htop</code></p>
<p><code>sshd: root@pts/0</code> means that there has been an SSH session established for the user <code>root</code> at the #<code>0</code> pseudoterminal (<code>pts</code>). A pseudoterminal emulates a real text terminal.</p>
<p><code>bash</code> is the shell that I am using.</p>
<p>Why is there a dash at the beginning? Reddit user hirnbrot helpfully explained it:</p>
<blockquote>
<p>There's a dash at the beginning because launching it as "-bash" will make it a login shell. A login shell is one whose first character of argument zero is a -, or one started with the --login option. This will then cause it to read a different set of configuration files.</p>
</blockquote>
<p><code>htop</code> is an interactive process viewer tool that is running in the screenshot.</p>
<p>After</p>
<pre><code>sudo apt remove lvm2 -y --purge sudo apt remove at -y --purge sudo apt remove snapd -y --purge sudo apt remove lxcfs -y --purge sudo apt remove mdadm -y --purge sudo apt remove open-iscsi -y --purge sudo apt remove accountsservice -y --purge sudo apt remove policykit-1 -y --purge </code></pre>
<p><img src="https://peteris.rocks/blog/htop/canyoukillit-after.png"></p>
<p>Extreme edition:</p>
<pre><code>sudo apt remove dbus -y --purge sudo apt remove rsyslog -y --purge sudo apt remove acpid -y --purge sudo systemctl stop cron &amp;&amp; sudo systemctl <span>disable</span> cron sudo rm /etc/systemd/system/getty.target.wants/getty@tty1.service sudo rm /lib/systemd/system/getty@.service </code></pre>
<p><img src="https://peteris.rocks/blog/htop/canyoukillit-after-extreme.png"></p>
<p>I followed the instructions in my blog post <a href="https://peteris.rocks/blog/unattended-installation-of-wordpress-on-ubuntu-server/">about unattended installation of WordPress on Ubuntu Server</a> and it works.</p>
<p>Here's nginx, PHP7 and MySQL.</p>
<p><img src="https://peteris.rocks/blog/htop/canyoukillit-after-extreme-wp.png"></p>
<p>Appendix</p>
<p>Source code</p>
<p>Sometimes looking at <code>strace</code> is not enough.</p>
<p>Another way to figure out what a program does is to look at its source code.</p>
<p>First, I need to find out where to start looking.</p>
<pre><code>$ <span>which</span> uptime /usr/bin/uptime $ dpkg -S /usr/bin/uptime procps: /usr/bin/uptime </code></pre>
<p>Here we find out that <code>uptime</code> is actually located at <code>/usr/bin/uptime</code> and that on Ubuntu it is part of the <code>procps</code> package.</p>
<p>You can then go to <a href="http://packages.ubuntu.com">packages.ubuntu.com</a> and search for the package there.</p>
<p>Here is the page for <code>procps</code>: <a href="http://packages.ubuntu.com/source/xenial/procps">http://packages.ubuntu.com/source/xenial/procps</a></p>
<p>If you scroll to the bottom of the page, you'll see links to the source code repositories:</p>
<p>File descriptors and redirection</p>
<p>When you want to redirect standard error (stderr) to standard output (stdout), is it <code>2&amp;&gt;1</code> or <code>2&gt;&amp;1</code>?</p>
<p>You can memorize where the ampersand <code>&amp;</code> goes by knowing that <code>echo something &gt; file</code> will write <code>something</code> to the file <code>file</code>. It's the same as <code>echo something 1&gt; file</code>. Now, <code>echo something 2&gt; file</code> will write the stderr output to <code>file</code>.</p>
<p>If you write <code>echo something 2&gt;1</code>, it means that you redirect stderr to a file with the name <code>1</code>. Add spaces to make it more clear: <code>echo something 2&gt; 1</code>.</p>
<p>If you add <code>&amp;</code> before <code>1</code>, it means that <code>1</code> is not a filename but the stream ID. So it's <code>echo something 2&gt;&amp;1</code>.</p>
<p>Colors in PuTTY</p>
<p><img src="https://peteris.rocks/blog/htop/putty.png"></p>
<p>If you have missing elements in htop when you are using PuTTY, here is how to solve it.</p>
<ul>
<li>Right click on the title bar</li>
<li>Click Change settings...</li>
<li>Go to Window -&gt; Colours</li>
<li>Select the Both radio button</li>
<li>Click Apply</li>
</ul>
<p><img src="https://peteris.rocks/blog/htop/putty-settings.png"></p>
<p>Shell in C</p>
<p>Let's write a very simple shell in C that demonstrates the use of <code>fork</code>/<code>exec</code>/<code>wait</code> system calls. Here's the program <code>shell.c</code>.</p>
<pre><code><span>#<span>include</span> <span>&lt;stdio.h&gt;</span></span> <span>#<span>include</span> <span>&lt;stdlib.h&gt;</span></span> <span>#<span>include</span> <span>&lt;unistd.h&gt;</span></span> <span>#<span>include</span> <span>&lt;string.h&gt;</span></span> <span>#<span>include</span> <span>&lt;sys/wait.h&gt;</span></span> <span><span>int</span> <span>main</span><span>()</span> </span>{ <span>printf</span>(<span>"Welcome to my shell\n"</span>); <span>char</span> line[<span>1024</span>]; <span>while</span> (<span>1</span>) { <span>printf</span>(<span>"&gt; "</span>); fgets(line, <span>sizeof</span>(line), <span>stdin</span>); line[<span>strlen</span>(line)<span>-1</span>] = <span>'\0'</span>; <span>// strip \n</span> <span>if</span> (<span>strcmp</span>(line, <span>"exit"</span>) == <span>0</span>) <span>// shell built-in</span> <span>break</span>; <span>int</span> pid = fork(); <span>if</span> (pid == <span>0</span>) { <span>printf</span>(<span>"Executing: %s\n"</span>, line); <span>if</span> (execlp(line, <span>""</span>, <span>NULL</span>) == <span>-1</span>) { <span>printf</span>(<span>"ERROR!\n"</span>); <span>exit</span>(<span>1</span>); } } <span>else</span> <span>if</span> (pid &gt; <span>0</span>) { <span>int</span> status; waitpid(pid, &amp;status, <span>0</span>); <span>printf</span>(<span>"Child exited with %d\n"</span>, WEXITSTATUS(status)); } <span>else</span> { <span>printf</span>(<span>"ERROR!\n"</span>); <span>break</span>; } } <span>return</span> <span>0</span>; } </code></pre>
<p>Compile the program.</p>
<pre><code>gcc shell.c -o shell </code></pre>
<p>And run it.</p>
<pre><code>$ ./shell Welcome to my shell &gt; date Executing: date Thu Dec 1 14:10:59 UTC 2016 Child exited with 0 &gt; <span>true</span> Executing: <span>true</span> Child exited with 0 &gt; <span>false</span> Executing: <span>false</span> Child exited with 1 &gt; <span>exit</span> </code></pre>
<p>Have you ever wondered that when you launch a process in the background you only see that it has exited only after a while when you hit <code>Enter</code>?</p>
<pre><code>$ sleep 1 &amp; [1] 11686 $ <span># press Enter</span> [1]+ Done sleep 1 </code></pre>
<p>That's because the shell is waiting for your input. Only when you enter a command does it check for the status of the background processes and show if they've been terminated.</p>
<p>TODO</p>
<p>Here is what I'd like to find out more about.</p>
<ul>
<li>process state substatuses (<code>Ss</code>, <code>Ss+</code>, <code>R+</code>, etc.)</li>
<li>kernel threads</li>
<li><code>/dev/pts</code></li>
<li>more about memory (<code>CODE</code>, <code>DATA</code>, <code>SWAP</code>)</li>
<li>figure out time slices length</li>
<li>Linux scheduler algorithm</li>
<li>pinning proceses to cores</li>
<li>write about manual pages</li>
<li>cpu/memory colors in bars</li>
<li>process ID limit &amp; fork bomb</li>
<li>
<code>lsof</code>, <code>ionice</code>, <code>schedtool</code>
</li>
</ul>
<p>Updates</p>
<p>Here is a list of non-minor corrections and updates since the post was published.</p>
<ul>
<li>Idle time in <code>/proc/uptime</code> is the sum of all cores (Dec 2, 2016)</li>
<li>My parent/child <code>printf</code> in <code>zombie.c</code> was reversed (Dec 2, 2016)</li>
<li>
<code>apt remove cron</code> installs <code>postfix</code> because of a dependency to an MTA (Dec 3, 2016)</li>
<li>
<code>id</code> can load information from other sources (via <code>/etc/nsswitch.conf</code>), not just <code>/etc/passwd</code> (Dec 3, 2016)</li>
<li>Describe <code>/etc/shadow</code> password hash format (Dec 3, 2016)</li>
<li>Use <code>visudo</code> to edit the <code>/etc/sudoers</code> file to be safe (Dec 3, 2016)</li>
<li>Explain <code>MEM%</code> (Dec 3, 2016)</li>
<li>Rewrite the section about load averages (Dec 4, 2016)</li>
<li>Fix: <code>kill 1234</code> by default sends <code>TERM</code> not <code>INT</code> (Dec 7, 2016)</li>
<li>Explain CPU and memory color bars (Dec 7, 2016)</li>
</ul>
<p>Final remarks</p>
<p>Please let me know if there is something wrong in this post! I will gladly correct it.</p>
<p>T-shirt</p>
<p>Congratulations on making to the end.</p>
<p>Here's another screenshot of htop.</p>
<p>This time it's of a human. It displays your inner processes and the load of your kernel.</p>
<p><a href="https://teespring.com/htop?ref=bottom-text">Get this t-shirt</a> for yourself or as a gift.</p>
<p>It's been bought 14 times already (including hoodies, etc.)</p>
<p><a href="https://teespring.com/htop?ref=bottom-img"><img src="https://peteris.rocks/blog/htop/htop-tshirt-terminal.jpg"></a></p>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via peteris.rocks<br>
<br>
November 1, 2019 at 06:06PM
